edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -154,6 +160,29 @@ bool TryMatchSubscription(SubscriptionPattern subscriptionPattern, string subscr
             return false;
         }
 
+        bool HasMixedIdentities(Option<string> deviceId, Option<string> moduleId, List<string> subscriptionList)
+        {
+            foreach (var subscriptionPattern in this.subscriptionPatterns)
+            {
+                foreach (var subscription in subscriptionList)
+                {
+                    var subscriptionMatch = Regex.Match(subscription, subscriptionPattern.Pattern);
```

Comment:
So -  this.subscriptionPatterns is the list of _current_ subscriptions for the client, and subscriptionList is the list of _new_ subscriptions for the same client, right?

If so, then why do you need to do a _match_ here? Can't you just check if _any_ of the old or new subscriptions are on-behalf-of?

Reply:
the subscriptionPatterns is a list of possible patterns for subscriptions, like for C2D or twin result. We don't know upfront that what we find in the list (the subscriptionList), so we need to try. The regex.Match is needed to obtain the identity from the subscription.
The situation here is the following: the broker said that device_x (or module_y on device_x) got a subscription change. So we have an identity (device_x). Next, we need to check that all the subscription device_x has right now (the subscribed topics in the subscriptionList), all have the identity of device_x. To do that, we need to know how to extract the identity from the subscription itself, that is what the matching does. It is a simple 'try until fits' approach. Once a pattern matches, we can get the device/module id from the subscription and then compare it to the identity we got as parameters.